lxd/device/gpu_sriov.go

```diff
@@ -83,25 +83,45 @@ func (d *gpuSRIOV) Start() (*deviceConfig.RunConfig, error) {
 	saveData := make(map[string]string)
 
 	// Get SRIOV parent, i.e. the actual GPU.
-	parentPCIAddress, err := d.getParentPCIAddress()
+	parentPCIAddresses, err := d.getParentPCIAddresses()
 	if err != nil {
 		return nil, err
 	}
 
-	// Get PCI information about the GPU device.
-	devicePath := filepath.Join("/sys/bus/pci/devices", parentPCIAddress)
+	var parentPCIAddress string
+	var pciParentDev pcidev.Device
+	var vfID int
+	var err1 error
```

Comment:
@monstermunchkin I'm not sure you need both these two err1 and err2 vars. From looking at the code it seems that err2 will only be set if err1 is nil, and if err1 is non-nil from the current iteration and is the last loop and err2 is non nil from the previous loop then it would return err1 anyway. 

So it seems like just using err would be fine. Unless I'm missing something here as to the reason why these errors need to be treated differently.

Reply:
@monstermunchkin also is it worth differentiating an "error" from "a GPU that has no free VF", i.e update `findFreeVirtualFunction` to not return an error if there are no free VFs, but use a -1 VF ID to indicate no free VF found (but no error while trying to find it). That way if a true unexpected error occurred you could return out of the loop directly.

Reply:
> So it seems like just using err would be fine. Unless I'm missing something here as to the reason why these errors need to be treated differently.

We could use a single `err` if we didn't care about where the error is coming from. I only used `err1` and `err2` to distinguish the errors.

> also is it worth differentiating an "error" from "a GPU that has no free VF", i.e update findFreeVirtualFunction to not return an error if there are no free VFs, but use a -1 VF ID to indicate no free VF found (but no error while trying to find it).

Absolutely. However, I'd use `-1` for errors and `0` for no free VFs.

Reply:
> That way if a true unexpected error occurred you could return out of the loop directly. And do the same for ParseUeventFile.

@tomponline I wouldn't do that. If the first GPU returned errors but a second one wouldn't, we'd fail for no reason.

Reply:
> Scratch that. `-1` makes total sense for no free VFs.

Yes I used that for the non-error scenario when failing to find any free VFs for NICs, as a valid VF can be 0.

Reply:
> > So it seems like just using err would be fine. Unless I'm missing something here as to the reason why these errors need to be treated differently.
> 
> We could use a single `err` if we didn't care about where the error is coming from. I only used `err1` and `err2` to distinguish the errors.
> 
> > also is it worth differentiating an "error" from "a GPU that has no free VF", i.e update findFreeVirtualFunction to not return an error if there are no free VFs, but use a -1 VF ID to indicate no free VF found (but no error while trying to find it).
> 
> Absolutely. However, I'd use `-1` for errors and `0` for no free VFs.

Ah so the two err vars are for differentiating between a failure in ParseUeventFile vs findFreeVirtualFunction so that you can errors.Wrapf() it with the correct context?

If so, why not do that in the associated `errN != nil` block, e.g.

```go
pciParentDev, err = pcidev.ParseUeventFile(filepath.Join(devicePath, "uevent"))
if err != nil {
    err = errors.Wrapf(err, "Failed to get PCI device info for GPU %q", parentPCIAddress)
    continue
}
```

Or even use a separate function for those two operations and return with the error context wrapped?